src/java/org/apache/cassandra/locator/ReplicaPlan.java

```diff
@@ -41,10 +45,11 @@
     //      ==> live.all()  (if consistencyLevel.isDCLocal(), then .filter(consistencyLevel.isLocal))
     private final E contacts;
 
-    ReplicaPlan(Keyspace keyspace, ConsistencyLevel consistencyLevel, E contacts)
+    ReplicaPlan(Keyspace keyspace, AbstractReplicationStrategy replicationStrategy, ConsistencyLevel consistencyLevel, E contacts)
```

Comment:
I wonder whether it would make sense to try to manage the strategy snapshotting inside the `ReplicaPlan`, instead of delegating it to the callers. We could try to do that with some additional per-implementation constructors doing the snapshot, keeping the ones with the `AbstractReplicationStrategy` argument private, [this way](https://github.com/adelapena/cassandra/commit/1b978976593c539097c06128264962555cc71ca7). wdyt?

Reply:
Some call-sites require to pass the `ReplicationStrategy` snapshot explicit, as it uses the snapshot to select the peers to contact, e.g. `org.apache.cassandra.locator.ReplicaPlans#forRead`
The other call-sites can safely and implicitly just pass the `keyspace` and get the latest RS in the constructor. 

One approach is to provide the both public constructor overrides. 

I tried it earlier, but it is confusing to users, IMO. Because it hides the side effect (i.e. calling the wrong constructor can lead to race). I'd favor explicitness. The client might need to write a bit more code, but easier for later maintenance.

Reply:
Fair enough :+1: